src/envoy/http/authn/authn_utils.cc

```diff
@@ -36,32 +36,41 @@ bool JsonIterCallback::JsonIteratorCallback(const std::string& key,
   ENVOY_LOG(debug, "{}: key is {}", __FUNCTION__, key);
   ::google::protobuf::Map< ::std::string, ::std::string>* claims =
       payload->mutable_claims();
+
+  // In current implementation, only string objects are extracted into
+  // claims. If call obj.asJsonString(), will get "panic: not reached" from
+  // json_loader.cc.
   try {
+    // Try as string, will throw execption if object type is not string.
+    std::string obj_str = obj.asString();
+    (*claims)[key] = obj_str;
+    // "aud" can be either string array or string.
+    // If it is string, save it to the claims and the payload
+    if (key == "aud") {
+      payload->add_audiences(obj_str);
+    }
+  } catch (Json::Exception& e) {
     if (key == "aud") {
       // "aud" can be either string array or string.
+      // If it is string array, only save it to the payload.
       // Try as string array, read it as empty array if doesn't exist.
-      std::vector<std::string> aud_vector;
       try {
-        aud_vector = obj.getStringArray(key, true);
-      } catch (Json::Exception& e) {
-        // Try as string
-        try {
-          auto audience = obj.asString();
-          aud_vector.push_back(audience);
-        } catch (Json::Exception& e) {
-          ENVOY_LOG(error, "aud field type is not string or string array");
+        std::string json_str = "{\"" + key + "\":" + obj.asJsonString() + "}";
```

Comment:
Please confirm this implementation with Wayne or someone (may be lookup cloudendpoints repo to see how it was done). Or leave a TODO (and create a tracking issue for this to remember) if you unsure. This new change doesn't feel right to me.

Reply:
Added the TODO. @qiwzhang : can you review the code mentioned in Diem's comment, which extracts the string array from aud?